### PR TITLE
Add From<i/u128> for JsBigInt

### PR DESCRIPTION
**Pull request by dranikpg**
This Pull Request fixes [1970](https://github.com/boa-dev/boa/issues/1970).

It changes the following:

- Added From<i128>, From<u128> for JsBigInt
- Added From<i128>, From<u128> for Numeric

### CODE REVIEW

**Review by Razican — APPROVED** [review]
Thank you for the contribution!! Looks great :)

**Review by jasonwilliams — APPROVED** [review]
Looks good to me

**Comment by Razican** [comment]
> Should i128 really be clamped to f64 like i64 is?
> 
> 
> 
> ```rust
> 
> impl From<i64> for JsValue {
> 
>     #[inline]
> 
>     fn from(value: i64) -> Self {
> 
>         if let Ok(value) = i32::try_from(value) {
> 
>             Self::Integer(value)
> 
>         } else {
> 
>             Self::Rational(value as f64)
> 
>         }
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> I'd rather expect something like this, as accuracy would be very low for big values otherwise
> 
> 
> 
> ```rust
> 
> impl From<i128> for JsValue {
> 
>     #[inline]
> 
>     fn from(value: i128) -> Self {
> 
>         if let Ok(value) = i32::try_from(value) {
> 
>             Self::Integer(value)
> 
>         } else {
> 
>             Self::BigInt(value.into()) // i.e. JsBigInt::from()
> 
>         }
> 
>     }    
> 
> }
> 
> ```

I would say this can be treated in a separate issue, it's a subjective topic :)

File: boa_engine/src/value/mod.rs [hunk 1]
```diff
@@ -1000,6 +1000,13 @@ impl From<f32> for Numeric {
     }
 }
 
+impl From<i128> for Numeric {
+    #[inline]
+    fn from(value: i128) -> Self {
+        Self::BigInt(value.into())
+    }
+}
+
```
Comment (HalidOdat) [inline boa_engine/src/value/mod.rs:1009]: This should be a number (`f64`) instead of BigInt

Comment (HalidOdat) [inline boa_engine/src/value/mod.rs:1009]: I'm no sure if we even want this Numeric is more of an internal thing.

Comment (Razican) [reply boa_engine/src/value/mod.rs:1009]: Well, we are already doing this with `i64`, and `BigInt` would make sure that it doesn't lose precision:
https://github.com/boa-dev/boa/blob/e2630faf82a27e958d0ca6c9b1c81554bcc07ff3/boa_engine/src/value/mod.rs#L1003-L1008

Comment (HalidOdat) [reply boa_engine/src/value/mod.rs:1009]: > Well, we are already doing this with i64

Ummm.... This is wrong, It should be a number, not bigint.

> BigInt would make sure that it doesn't lose precision

It's fine if we lose precision (we lose precision if we type a number value which is too big for `f64` in JavaScript, same with `From<i64>` for `JsValue`), If we don't want loss of precision we create a bigint and we can convert it into `Numeric`.
The behaviour of `From<u/iN>` should be uniform, to do some number and others bigint is confuzing.

Comment (Razican) [reply boa_engine/src/value/mod.rs:1009]: There was some discussion here too: #1961

From my point of view, we should be able to convert an `f64` to `JsValue` or `Numeric`, and a `BigInt` too. But nothing else (at most an i32). If the user wants to convert a `i128` to a `BigInt`, they can first convert it to a `BigInt` and then convert the `BigInt` to a `Numeric` or a `JsValue`.

We shouldn't probably force anything, but give some options. Maybe a `TryFrom` implementation could be nice.

Comment (HalidOdat) [reply boa_engine/src/value/mod.rs:1009]: > [](https://github.com/boa-dev/boa/issues/1961)
> 
> From my point of view, we should be able to convert an `f64` to `JsValue` or `Numeric`, and a `BigInt` too. But nothing else (at most an i32). If the user wants to convert a `i128` to a `BigInt`, they can first convert it to a `BigInt` and then convert the `BigInt` to a `Numeric` or a `JsValue`.
> 
> We shouldn't probably force anything, but give some options. Maybe a `TryFrom` implementation could be nice.

Yup. agreed

File: boa_engine/src/value/mod.rs [hunk 2]
```diff
@@ -1000,6 +1000,13 @@ impl From<f32> for Numeric {
     }
 }
 
+impl From<i128> for Numeric {
+    #[inline]
+    fn from(value: i128) -> Self {
+        Self::BigInt(value.into())
+    }
+}
```
Comment (HalidOdat) [inline boa_engine/src/value/mod.rs:1008]: We should remove this, as well as `From<u128> for Numeric`. If the user wants to a bigint or number they should explicitly say so.

Besides it's out of the scope of this PR of adding `From<i/u128> for JsBigInt`

**Comment by lastmjs** [comment]
Don't we want to implement the `From<i128>` trait for `JsValue`? Why aren't we adding an implementation in here: https://github.com/boa-dev/boa/blob/main/boa_engine/src/value/conversions.rs?

**Comment by Razican** [comment]
This will be handled as part of #2276.

**Pull request closed without merging**